src/pcmp.rs

```diff
@@ -130,7 +152,7 @@ unsafe fn first_start_of_match_inner(text: &[u8], pat: &[u8], p: __m128i) -> Opt
     }
     while text.len() >= offset - tp_align_offset + patl {
         let tlen = text.len() - (offset - tp_align_offset);
-        let ret = pcmpestri_16(tp_aligned, offset, tlen, p, patl) as usize;
+        let ret = pcmpestri_16_mask(tp_aligned, offset, tlen, p, patl) as usize;
```

Comment:
Btw, this is one missed optimisation that I didn't use in my branch: both regular intrinsic and our fallback assume pointer to be arbitrarily aligned, so they have to create a backing storage and use pointer copy to safely initialise SIMD vector. But for places like this one where we know that pointer is already properly aligned and sized for a SIMD instruction, we could be more efficient by simply casting it to `*const __m128i` and dereferencing.

Reply:
I think we still don't know if all 16 bytes past the pointer are readable, so we can't?

The code that does pointer alignment seems to be unnecessary? We don't really care for that in the pcmpestri instructions we run.

Reply:
I pushed a change to remove unnecessary alignment code (no need before the masked load as said). Should clear up some confusing parts of the code. On the other hand, the first_start_of_match_nomask could maybe use alignment, if it is any improvement to switch to aligned loads. I think on new arches it doesn't make much difference?

Reply:
Didn't the loop execute for a checked length >= 16 only anyway? If not, that can be easily added. I'd expect that path with aligned loads to be faster when it's in a hot path (for large needles), but for common short ones it won't make a difference.

Reply:
Yes there's main loops for where the length is >= 16 and then loops for the tail. Current code is maybe way overcomplicated by using pcmpestri even for the tail?

Looking at that revealed that we often scan twice (both in the main loop and then in the tail loop), since we forget to update the position if nothing was found in the first loop. I'm fixing that too, should be a sizable win for the case where we don't have matches..